File: permutation-in-string/Solution567.java
// https://leetcode.com/problems/permutation-in-string/

import java.util.*;

class Solution567 {

    // T: O(n), S: O(1)
    public boolean checkInclusion(String s1, String s2) {
        int n1 = s1.length(), n2 = s2.length();
        var counts1 = new int[26];
        var counts2 = new int[26];
        for (int i = 0; i < n1; i++) {
            counts1[s1.charAt(i) - 'a']++;
            counts2[s2.charAt(i) - 'a']++;
        }
        for (int i = n1; i < n2; i++) {
            if (Arrays.equals(counts1, counts2)) {
                return true;
            }
            counts2[s2.charAt(i - n1) - 'a']--;
            counts2[s2.charAt(i) - 'a']++;
        }
        return Arrays.equals(counts1, counts2);
    }

    public static void main(String[] args) {
        var sol = new Solution567();
        System.out.println(sol.checkInclusion("ab", "eidbaooo")); // true
        System.out.println(sol.checkInclusion("ab", "eidboaoo")); // false
        System.out.println(sol.checkInclusion("adc", "dcda")); // true
    }

}

File: permutation-in-string/solution567.py
# https://leetcode.com/problems/permutation-in-string/

class Solution567:

    # T: O(n), S: O(1)
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n1, n2 = len(s1), len(s2)
        if n1 > n2:
            return False

        counts1 = [0] * 26
        counts2 = [0] * 26

        for i in range(n1):
            counts1[ord(s1[i]) - ord('a')] += 1
            counts2[ord(s2[i]) - ord('a')] += 1

        for i in range(n1, n2):
            if counts1 == counts2:
                return True
            counts2[ord(s2[i - n1]) - ord('a')] -= 1
            counts2[ord(s2[i]) - ord('a')] += 1

        return counts1 == counts2


def main():
    sol = Solution567()
    print(sol.checkInclusion("ab", "eidbaooo"))  # True
    print(sol.checkInclusion("ab", "eidboaoo"))  # False
    print(sol.checkInclusion("adc", "dcda"))     # True


if __name__ == "__main__":
    main()
